### PR TITLE
cap[agentic-grpo/pi-compaction]: framework + iter 0 baseline + iter 1 rollouts + long-context training finding

### PR DESCRIPTION
**Pull request by ericflo**
## What

Second capability in the agentic-pi GRPO series. Sister cap to the
just-merged `pi-doctest` (PR #1051). This one targets **pi compaction**:
the structured-summary task pi runs when a long session exceeds its
context budget.

Pi spec being targeted: `packages/coding-agent/src/core/compaction/`
in `earendil-works/pi` — specifically `compaction.ts` +
`utils.ts::serializeConversation` + `SUMMARIZATION_PROMPT`. The rollout
input matches what production pi sends the model byte-for-byte.

## Status

**Framework + baseline + iter 1 rollouts shipped.** No trained adapter
yet — kiln-train on pi-compaction-sized GRPO groups proved
prohibitively slow on the H100 pod within the lease TTL. Next iter
needs either a kiln-train long-context optimization or a much
smaller-window training approach.

## Iter 0 — Baseline established

Clean baseline (n=19 successful rollouts out of 23; 4 hit network errors
from a pod-sharing conflict — see kiln-polish.jsonl):

| metric | value |
|---|---|
| mean composite | 0.328 ± 0.346 |
| pass-gate (composite > 0.5) | 7/19 |
| fail-gate (~0.07) | 12/19 |
| distribution | bimodal (~0.77 cluster vs ~0.07 cluster) |

The base Qwen3.5-4B emits pi-format on **~37% of tasks** unprompted —
strong gradient signal for GRPO to bias toward the format-passing region.

## Iter 1 — Training rollouts collected, GRPO step too slow

Iter 1c with source truncated to last 30K chars:

- 48 training rollouts on 12 short tasks (each task gets 4 generations)
- Mean composite 0.558 on truncated input (vs 0.328 on full-source baseline)
- **4 of 12 groups pass strong-signal filter** (var > 0.05) — these
  are bimodal `[0.07, pass, pass, pass]` tasks where GRPO would have
  strong gradient
- 6 already-pass groups (~0.78) and 2 always-fail groups (~0.07)
  dropped by the filter
- GRPO step on the 4 strong-signal groups failed to fit in pod TTL:
  no progress= line in 6 min with KILN_GRAD_CHECKPOINT_SEGMENTS=8
  (pi-doctest training took 140s total for 11 groups — pi-compaction
  has ~10× more tokens per group due to long source_text)

This is captured as a kiln-polish entry (training-too-slow-on-long-context-grpo).

## The world-class rubric (v1.2)

Eric's mandate was "spend just an enormous amount of thought effort and
energy making this the best you can possibly make it" — so the rubric is
intentionally multi-component with anti-shortcut counters paired to every
reward-presence sub-score:

```
composite = (
    0.10 * format_score        # rigid pi-format compliance (8 sub-checks)
  + 0.30 * content_score        # preserves source facts (8 sub-checks incl. recency)
  + 0.20 * faithfulness_score   # no hallucination (reverse-direction, 4 sub-checks)
  + 0.10 * compression_score    # smaller + ratio band (5%-25% ideal)
  + 0.10 * continuability_score # next-steps + in-progress concrete
  + 0.20 * outcome_gate         # hard gate: format >= 0.5 AND content >= 0.30 AND faithfulness_paths >= 0.7
)
```

Recency-weighted recall (v1.2): the most load-bearing facts in a
compaction summary are the *recent* ones — pi will literally pick up
from the last few turns. Three sub-scores target this:
`content.recent_paths_recall`, `content.recent_identifiers_recall`,
`content.last_actions_recall`.

**Calibration PASSES** (good_mean=0.827, bad_max=0.055):

| shortcut | mean | counter |
|---|---|---|
| continuation | 0.026 | format.no_continuation |
| copy_source | 0.046 | compression + content |
| hallucinated | 0.052 | faithfulness.file_paths_in_source |
| template_only | 0.053 | content_score gate |
| wrong_format | 0.034 | format.* |

15× headroom between good and bad. Adversarial design audit in
capability.md walks through each cheap-shortcut path and how the rubric
counters it.

## What's in this PR

- `capability.md` — design + adversarial audit + rubric weighting plan
- `rubric.py` (v1.2) — multi-component scorer (688 lines)
- `task_scaffold.py` — pi serializeConversation mirror + ground-truth
  extractor (recent paths/identifiers + last_actions)
- `build_corpus.py` — samples long Claude conversations 30K-120K tokens
  from trajectories.db, length-stratified, **with secret scrubber** for
  RunPod / GitHub / OpenAI / Anthropic / AWS keys
- `build_calibration.py` — programmatic gold-summary generator + 5
  shortcut shapes (continuation, copy_source, hallucinated, template,
  wrong_format)
- `rubric_sanity.py` — gates good_mean ≥ 0.55, bad_mean ≤ 0.30,
  good_min ≥ bad_max - 0.05
- `rollout.py` — kiln /v1/chat/completions rollout harness
- `backup_to_b2.py` — pulls adapter + rollouts + summaries from the
  pod, uploads to b2://clouderic/kiln/pi-compaction/<date>/iter-<n>-<kind>/
  with sha256-checked manifest
- `run_iter.sh` / `drive_iters.sh` — autonomous iter loop scaffolding
- `truncate_corpus.py` — produces source_text-truncated copies for
  tractable training
- `kiln-polish.jsonl` — two durable findings:
  1. **pool-isolation**: kiln-pool pods are shared mutable state. An
     unrelated autonomous loop on the same pod restarted kiln serve
     mid-iter0-eval, causing 4/23 tasks to hit RemoteDisconnected.
  2. **training-too-slow-on-long-context-grpo**: cuda_grpo_ablation
     on pi-compaction-sized groups (~30K tokens each) doesn't make
     visible progress in 6 min even with KILN_GRAD_CHECKPOINT_SEGMENTS=8.
     pi-doctest training (~3K tokens per group) completes in 2-3 min
     on the same hardware. The 10× slowdown is super-linear in token
     count.
- `iter-artifacts/` — iter 0 + iter 1 rollouts.jsonl + summary.json
  for downstream analysis

## B2 backup

Every iter's artifacts are pushed to
`b2://clouderic/kiln/pi-compaction/<YYYYMMDD>/iter-<N>-<kind>/`
with sha256-checked manifest. The framework + iter 0 baseline + iter 1
rollouts are durably backed up:

- `iter-0-baseline/` — base model eval rollouts (n=23)
- `iter-1-rollouts-only/` — training rollouts (n=48, 12 tasks × 4 gens,
  truncated to 30K chars source) + capability source snapshot

## Next steps

When pod time + a kiln-train long-context optimization become available:

1. Re-run iter 1 GRPO step on the existing 4-strong-signal rollouts (the
   data is checked into iter-artifacts/iter1c-rollouts.rollouts.jsonl)
2. Eval the resulting adapter against the held-out eval set (truncated
   to 30K chars to match training input distribution)
3. Compare to iter 0 baseline 0.328 — falsification threshold +0.05
   composite
4. Iter 2: H12 strong-signal filter with re-rolled training data
5. Iter 3+: standard pi-doctest GRPO discipline (multi-seed
   verification, watch wall-clock as a regression metric)

The rubric is robust enough that any future adapter can be plugged in
and scored with no additional setup.

## Pinned references

- pi compaction source:
  https://github.com/earendil-works/pi/tree/main/packages/coding-agent/src/core/compaction
- prior cap (merged): PR #1051 (pi-doctest, 3-seed-verified +4.2pp mean)
- skill: `.agents/skills/agentic-grpo-capability-creator/SKILL.md`
- B2 artifacts: `b2://clouderic/kiln/pi-compaction/20260518/`

### CODE REVIEW

**Comment by ericflo** [comment]
## Iter 2 update (2026-05-19)

**Landed**: kiln-train shared prompt-prefix reference forward (#1055, on main). Cuts GRPO ref-forward attention work `~n_completions×` for long-context groups by forwarding the shared prompt once via the paged path, snapshotting GDN linear state, and forwarding only each completion's tokens with `start_pos = prompt_len`.

**Iter 2 work** (this branch):

- **Full-length rollout pipeline validated**. 8 training tasks × 4 generations on FULL pi-serialized inputs (no truncation): 32/32 rollouts, mean composite 0.570, 3/8 strong-signal groups (var > 0.05). Wall clock 69s mean per rollout.
- **Baseline replicates**. 23-task eval with base model: composite 0.314 (vs iter 0's 0.328 — Δ -0.014, well within noise). Bimodal preserved.
- **No trained adapter yet.** The cuda_grpo_ablation GRPO step does not complete on the strong-signal groups within practical wall-clock budgets at long context (>7K-token sequences). Two failure modes:
  - Default env (`KILN_GRAD_CHECKPOINT_SEGMENTS=8`, `KILN_STREAMING_TILE_TOKENS=2048`): OOM at streaming GDN tile 2 of 4 at layer 30 (`DriverError CUDA_OOM`).
  - Aggressive env (`KILN_GRAD_CHECKPOINT_SEGMENTS=32`, `KILN_STREAMING_TILE_TOKENS=512`, `KILN_SPOOL_CHECKPOINT_BOUNDARIES=1`): VRAM fits (12 GB used) but CPU-bound at 99% on kernel-launch overhead from many small GDN streaming tiles — no progress= line in 10 min on 3 groups.

**Next bottleneck** (logged as iter 2 kiln-polish entry `long-context-policy-forward-bottleneck`): the streaming GDN per-tile transient + grad-checkpoint recompute. Likely needs either (a) tile-internal checkpointing of GDN intermediates so larger tiles can fire without OOM, or (b) fused multi-layer chunked GDN+attn kernels that amortize launch overhead across layers.

**B2 backup**: `b2://clouderic/kiln/pi-compaction/20260519/iter-2-train/` (manifest + rollouts + summaries + scrubbed corpus).

**Iter 2 artifacts** committed to this branch under `iter-artifacts/iter2-*.jsonl|.summary.json`. The strong-signal training data is ready for downstream training as soon as the policy-forward bottleneck is unblocked.
